File: runtime.py
import collections
import pickledb
import logging

class Transaction(object):
    def __init__(self):
        pass
    def commit(self, dependencies, data):
        pass
class Runtime(object):
    def __init__(self, db, upcalls):
        self.db = db
        self.local_horizon = collections.defaultdict(lambda : 0)
        self.upcalls = upcalls
    def global_horizon(self, id):
        h = self.get(str(id) + ':' + 'count')
        if h:
            return h
        else:
            return 0
    def set(self, key, value):
        return self.db.set(str(hash(key)), value)
        # return self.db.set(hash(key), value)
    def set_entry(self, id, index, value):
        return self.set(str(id) + ':' + str(index), value)
    def get(self, key):
        # return self.db.get(hash(key))
        return self.db.get(str(hash(key)))
    def get_entry(self, id, index):
        return self.get(str(id) + ':' + str(index))
    def read(self, id, index):
        # print id, index, self.local_horizon[id]
        if self.local_horizon[id] >= index:
            return None
        # follow dependencies
        val = self.get_entry(id, index)
        # print val
        if val != None:
            for n_id, n_index in val[0]:
                self.read(n_id, n_index)
            self.local_horizon[id] += 1
            # upcall
            logging.info('Make upcall ' + str(id) + ' with arg ' + str(val[1]))
            self.upcalls[id](val[1])
        else:
            return None
# API
    def append(self, id, dependencies, data):
    # id is a column id
    # dependencies are [(id, index)]
    # data are of the datatype
    # return an index
        horizon = self.global_horizon(id)
        # print "(dependencies, data)", (dependencies, data)
        self.set_entry(id, horizon + 1, (dependencies, data))
        # Check if false?
        self.set(str(id) + ':' + 'count', horizon + 1)
        self.db.dump()
        return horizon + 1
    def read_next(self, id):
        seen = self.local_horizon[id]
        return self.read(id, seen + 1)
    def play_forward(self, id):
        while self.local_horizon[id] < self.global_horizon(id):
            self.read_next(id)
    def start_transaction(self, ids):
        '''
        Read forward the id's
          if all transactions are committed
          then add transaction start entries to the ids, containing the min id
        '''
        return Transaction()

# id : (index, dependencies, data)
# id+

File: runtime.cpp
#include <iostream>
#include <string>
#include <type_traits>
#include <vector>
#include <map>
#include <string>

using Dependency = std::pair<int, int>;
using DependencyList = std::vector<Dependency>;
using Data = int;
using Id = int;
using Index = int;
using NodeValue = std::pair<DependencyList, Data>;
using NodeValuePtr = std::shared_ptr<NodeValue>;
class Database {
    public :
        Database();
        NodeValuePtr get(Id id, Index index);
        int set(Id id, Index index, const NodeValue& node_value);
        Index get_global_horizon(Id id);
        Index set_global_horizon(Id id, Index new_horizon);
    private :
        std::map<std::string, NodeValue> _map;
        std::map<std::string, Index> _g_horizon;
};

class BaseLocalView {
    public :
        BaseLocalView();
        virtual int upcall(Id id, Data data);
};

class Runtime {
    public :
        Runtime(Database& db, BaseLocalView& local_view);
        int append(Id id, DependencyList dependencies, Data data);
        int read_next(Id id);
        int play_forward(Id id);
    private :
        Index get_local_horizon(Id id);
        Index set_local_horizon(Id id, Index new_horizon);
        int read(Id id, int index);
        Database& _db;
        BaseLocalView& _local_view;
        std::map<std::string, Index> _local_horizon;
};

Database::Database() {
}
NodeValuePtr Database::get(Id id, Index index) {
    auto id_str = std::to_string(id);
    auto index_str  = std::to_string(index);
    auto key = id_str + ":" + index_str;
    auto it = _map.find(key);
    if (it == _map.end()) {
        return std::shared_ptr<NodeValue>(0);
    } else {
        return std::shared_ptr<NodeValue>(&(it->second));
    }
}
int Database::set(Id id, Index index, const NodeValue& node_value) {
    auto id_str = std::to_string(id);
    auto index_str  = std::to_string(index);
    auto key = id_str + ":" + index_str;
    _map.insert(std::pair<std::string, NodeValue>(key, node_value));
    return 0;
}
Index Database::get_global_horizon(Id id) {
    auto it  = _g_horizon.find(std::to_string(id) + ":count");
    if (it == _g_horizon.end()) {
        return 0;
    } else {
        return it->second;
    }
}
Index Database::set_global_horizon(Id id, Index new_horizon) {
    _g_horizon.insert(std::pair<std::string, Index>(
                std::to_string(id) + ":count",
                new_horizon));
    return new_horizon;
}

Runtime::Runtime(Database& db, BaseLocalView& local_view) 
    : _db(db), _local_view(local_view) {
}
Index Runtime::get_local_horizon(Id id) {
    auto it  = _local_horizon.find(std::to_string(id) + ":count");
    if (it == _local_horizon.end()) {
        return 0;
    } else {
        return it->second;
    }
}
Index Runtime::set_local_horizon(Id id, Index new_horizon) {
    _local_horizon.insert(std::pair<std::string, Index>(
                std::to_string(id) + ":count",
                new_horizon));
    return new_horizon;
}
int Runtime::append(Id id, DependencyList dependencies, Data data) {
    auto horizion = _db.get_global_horizon(id);
    _db.set(id, horizion + 1, NodeValue(dependencies, data));
    _db.set_global_horizon(id, horizion + 1);
    return 0;
}
int Runtime::read_next(Id id) {
    auto seen = get_local_horizon(id);
    return read(id, seen + 1);
}
int Runtime::read(Id id, Index index) {
    if (get_local_horizon(id) >= index) {
        return -1;
    }
    auto val = _db.get(id, index);
    if (val) {
        
    } else {
        return -1;
    }

    return 0;

}








    
    



int main() {
    std::cout << "hi";
}
